QuantumKernel refactor

adekusar-drl opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
No new features, the `QuantumKernel` code is just split into smaller methods.

dlasecki commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -355,22 +357,27 @@ def construct_circuit(
-                "x and class feature map incompatible dimensions.\n"
-                f"x has {len(x)} dimensions, but feature map has {self._feature_map.num_parameters}."
-            )
+        if y is None:
+            y = x
```
```suggestion
        y = x if y is None
```
For compactness.

adekusar-drl replied:
`else` is required :)
could be `y = x if y is None else y` but I think it is better.

dlasecki replied:
You are right, my bad.

dlasecki commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -380,38 +387,34 @@ def construct_circuit(
 
-            if y is None:
-                y = x
+    # just a synonym
+    def _construct_circuit_statevector(self, x: ParameterVector, y: ParameterVector = None):
+        """This is just a synonym for ``_construct_circuit_with_feature_map``"""
+        return self._construct_circuit_with_feature_map(x, y)
```
Why do we need a synonym?

adekusar-drl replied:
This is a good question, there's a bunch of methods like `whatever()` and `whatever_statevector()` now. But `_construct_circuit_with_feature_map()` is used to construct circuits in both cases: statevector and non-statevector. That's why it is named like now. I can rename it to `_construct_circuit_statevector()`, but then it would be used in `construct_circuit()` that is used in the non-statevector case and such code might be confusing. I hope I did not bring more confusion.

gentinettagian replied:
I think in this case it would be less confusing to just have one `_construct_circuit_with_feature_map()` method that is used for both statevector and hardware/qasm-simulator. At least when glancing over the code this synonym method confused me at first. At the places where the logic differs it is definitely nice to separate the code into separate methods.

adekusar-drl replied:
@gentinettagian I'm not saying I'm against the change, I wonder what is exactly confusing as method name consistency is also important.

dlasecki commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -380,38 +387,34 @@ def construct_circuit(
+        Helper function to compute overlap for given input if a statevector simulator is used.
+        """
+        # |<0|Psi^dagger(y) x Psi(x)|0>|^2, take the amplitude
+        v_a, v_b = [results[int(i)] for i in idx]
```
Is the typhint for idx correct? Casting a tuple onto int would not work...

adekusar-drl replied:
The type hint was correct! I understand the code was confusing, I rewrote it without tuples, I hope now it is easier to understand.

dlasecki commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -441,188 +444,224 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+        if is_symmetric:
+            np.fill_diagonal(kernel, 1)
+
+        is_statevector_sim = self._quantum_instance.is_statevector
```
How about we skip this assignment? The assigned variable is used only once.

adekusar-drl replied:
Makes sense. Removed the variable.

dlasecki commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -441,188 +444,224 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+
+        return kernel
+
+    def _validate_input(self, x_vec: np.ndarray, y_vec: np.ndarray):
```
Return type hint missing.

adekusar-drl replied:
Added type hints.

dlasecki commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -441,188 +444,224 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+
+        return kernel
+
+    def _validate_input(self, x_vec: np.ndarray, y_vec: np.ndarray):
```
Validate might suggest that vectors are not modified in the method but they might be.

dlasecki commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -441,188 +444,224 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+        Args:
+            vec: A vector to validate.
+            vec_name: A vector name, if an error is raised it ise used in the error message.
+            adjust_feature_map: If ``True`` tried to adjust the number of qubits of the feature map.
```
```suggestion
            adjust_feature_map: If ``True`` tries to adjust the number of qubits of the feature map.
```

adekusar-drl replied:
This method is removed now.

dlasecki commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -441,188 +444,224 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+            if adjust_feature_map:
+                try:
+                    # an attempt to update the feature map with the required number of qubits
+                    self._feature_map.num_qubits = vec
```
Why do we assign a vector to a field holding an int?

adekusar-drl replied:
Yeah, that was wrong. Anyway, the method is removed.

dlasecki commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -441,188 +444,224 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+
+    def _calculate_kernel_statevector(
+        self, x_vec: np.ndarray, y_vec: np.ndarray, is_symmetric: bool, kernel: np.ndarray
+    ):
```
Return type hint missing. Please check other places.

adekusar-drl replied:
Reviewed type hints here and there.

dlasecki commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -441,188 +444,224 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+        parameterized_circuit = self._quantum_instance.transpile(
+            parameterized_circuit, pass_manager=self._quantum_instance.unbound_pass_manager
+        )[0]
+        statevectors = []
+
+        for min_idx in range(0, len(to_be_computed_data), self._batch_size):
+            max_idx = min(min_idx + self._batch_size, len(to_be_computed_data))
+            circuits = [
+                parameterized_circuit.assign_parameters({feature_map_params: x})
+                for x in to_be_computed_data[min_idx:max_idx]
+            ]
+            if self._quantum_instance.bound_pass_manager is not None:
+                circuits = self._quantum_instance.transpile(
+                    circuits, pass_manager=self._quantum_instance.bound_pass_manager
+                )
+            results = self._quantum_instance.execute(circuits, had_transpiled=True)
+            for j in range(max_idx - min_idx):
+                statevectors.append(results.get_statevector(j))
```
I would extract it to a method.

adekusar-drl replied:
I'd prefer to keep as it's now for consistency with `_calculate_kernel()`.

dlasecki reviewed (changes requested):
Performed the first scan and left some comments.
Overall, the class is very lengthy. Is there a way to break it down into smaller classes with clear and focused responsibilities?

adekusar-drl commented:
@gentinettagian Thanks for looking into this and your comment are very welcome! I'll try to think what can be improved further, but considering a new incoming primitive based version, I don't want to spend much time on this one. Perhaps, splitting into methods should be enough :)

woodsp-ibm commented on qiskit_machine_learning/kernels/quantum_kernel.py:
```diff
@@ -441,188 +453,208 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+            )
 
-        return kernel
+    def _raise_incompatible_feature_map(self, vec_name: str, vec_len: int) -> None:
```
The return type here ought to be `NoReturn` I think since all this function never ever returns

woodsp-ibm commented:
Just a comment. I will note that overall we could do better with testing error condition paths. Here with it refactored as a method that ends up not being tested at all - not that its really any worse than before. Something to improve upon with the unit testing perhaps when we migrate these over to using primitives.

adekusar-drl commented:
> Just a comment. I will note that overall we could do better with testing error condition paths. Here with it refactored as a method that ends up not being tested at all - not that its really any worse than before. Something to improve upon with the unit testing perhaps when we migrate these over to using primitives.

Added some more tests. Removed one of the `if` statements that always evaluates to true.

adekusar-drl commented:
@woodsp-ibm please review, last updates were just merges.
